extensions/http-curl/client/HTTPClient.cpp

```diff
@@ -328,18 +325,15 @@ void HTTPClient::setRequestHeader(std::string key, std::optional<std::string> va
 }
 
 namespace {
-struct CurlSListFreeAll {
-  void operator()(struct curl_slist* slist) const {
-    curl_slist_free_all(slist);
-  }
-};
+using CurlSlistDeleter = decltype([](struct curl_slist* slist) { curl_slist_free_all(slist); });
 
-std::unique_ptr<struct curl_slist, CurlSListFreeAll> getCurlSList(const std::unordered_map<std::string, std::string>& request_headers) {
+std::unique_ptr<struct curl_slist, CurlSlistDeleter> toCurlSlist(const std::unordered_map<std::string, std::string>& request_headers) {
   curl_slist* new_list = nullptr;
+  const auto guard = gsl::finally([&new_list]() { curl_slist_free_all(new_list); });
   for (const auto& [header_key, header_value] : request_headers)
     new_list = curl_slist_append(new_list, utils::StringUtils::join_pack(header_key, ": ", header_value).c_str());
```

Comment:
we could eliminate the need for the guard and the raw pointer, we also did not handle the error case of `curl_slist_append` returning a nullptr

from [curl.se](https://curl.se/libcurl/c/curl_slist_append.html)
`A null pointer is returned if anything went wrong, otherwise the new list pointer is returned. To avoid overwriting an existing non-empty list on failure, the new list should be returned to a temporary variable which can be tested for NULL before updating the original list pointer. `

```suggestion
  for (const auto& [header_key, header_value] : request_headers)
    std::unique_ptr<struct curl_slist, CurlSlistDeleter> appended_list{curl_slist_append(new_list.get(), utils::StringUtils::join_pack(header_key, ": ", header_value).c_str())};
    if (!appended_list) {
      throw std::runtime_error("Failed to append to slist");
    }
    new_list.release();
    new_list = std::move(appended_list);
```

Reply:
Handled the possible failure of `curl_slist_append` in [ef82531](https://github.com/apache/nifi-minifi-cpp/pull/1648/commits/ef82531959f1ee395775a825db37ccb8d68daf47).

I used a different approach that avoids overlapping ownership and the calling of `release`, because both of these make me uncomfortable. The raw pointer remained, but annotated it with `gsl::owner`. I should have done that originally.